src/processors/default/processTrackArtworks.ts

```diff
@@ -0,0 +1,115 @@
+import { readFile, writeFile, rm } from 'node:fs/promises';
+import http from 'node:http';
+import https from 'node:https';
+import Path from 'node:path';
+
+import im from 'imagemagick';
+import { create } from 'ipfs-http-client';
+import tempDirectory from 'temp-dir';
+
+import { Table } from '../../db/db';
+import { missingProcessedArtworks } from '../../triggers/missing';
+import { Clients } from '../../types/processor';
+import { ProcessedTrack } from '../../types/track';
+import { rollPromises } from '../../utils/rollingPromises';
+
+const name = 'addMetadataIPFSHash';
+
+function imageBuffer(url: string): Promise<Buffer> {
+  console.log(`Downloading ${url}`)
+  return new Promise(function (resolve, reject) {
+    const client = url.includes('https') ? https : http;
+    client.get(url, function (res) {
+      const data: any = []
+      res.on('data', function (chunk) {
+        data.push(chunk);
+      }).on('end', function () {
+        const buffer = Buffer.concat(data);
+        resolve(buffer);
+      }).on('error', function (e) {
+        reject(e)
+      });
+    });
+  })
+}
+
+async function resizeImage(originalPath: string, size: string): Promise<Buffer> {
+  let resizedBuffer = null;
+  const buffer = await withTempPath(function (outPath) {
+    console.log(`resizing image to ${size} ${outPath}`)
+    return new Promise(function (resolve, reject) {
+      im.convert([originalPath, '-resize', size, outPath],
+        async function (err, stdout) {
+          if (err) reject(err);
+          resizedBuffer = await readFile(outPath)
+          resolve(resizedBuffer);
+        });
+    })
+  }, 'png');
+  return buffer;
+}
+
+async function uploadBuffer(buffer: Buffer) {
+  const url: any = '/ip4/127.0.0.1/tcp/5011'; // have to use any because create only accepts hash
+  const client = create(url)
+  const { cid } = await client.add(buffer)
+  return cid;
+}
+
+async function withTempPath(callback: (path: string) => Promise<any>, ext: string | void): Promise<any> {
+  const fileName = (Math.random().toString(16) + '0000000').substr(2, 8);
+  const path = `${Path.join(tempDirectory, fileName)}${ext ? '.' + ext : ''}`;
+
+  try {
+    return await callback(path);
+  } catch (e) {
+    throw e;
+  } finally {
+    await rm(path);
+  }
+}
+
+async function temporaryWrite(buffer: Buffer, callback: (path: string) => Promise<void | Buffer>): Promise<void> {
+  return withTempPath(async function (path) {
+    await writeFile(path, buffer)
+    await callback(path);
+  })
+}
+// when i use ProcessedTrack it claims lossyArtworkURL does not exist :shrug:
```

Comment:
yeh - its an optional field:

```
type ProcessedTrackArtwork = { lossyArtworkURL: string } | { lossyArtworkIPFSHash: string };
```

So the type claims that at least one of these fields will be there, but not necessarily both. E.g. There are some music nft platforms that only provide centralized artwork URL and no ipfs involvement, and there are other ones that only provide IPFS hash and no URL.

I think IPFS hash should take priority, so logic should be someth like:
 - If is IPFS hash, get image from ipfsClient.getHTTPURL(processedTrack.lossyArtworkIPFSHash);
 - If no IPFS hash, get image from processedTrack.lossyArtworkURL

Also, for the first case in the condition above to actually work, we'll require:
 - lossyArtworkIPFSHash has been successfully pinned
 - pinning has completed,

otherwise there is a race condition where this processTrackArtwork runs before the ipfsHash has completed pinning. anyway i think for now its fine, don't worry about this and if the race condition occurs it will just error and can clean errors+retry later
 


as the URL may not exist if for example its only on IPFS, in which case need to pull it from our IPFS gateway as in src/clients/ipfs.ts:getHTTPURL
 - hm this also means that pinning the

Reply:
ipfsHash first prio is coz the hash is stored on chain and so more robust. whether the actual image is there or not is another question, but yes i think also more likely that the image exists too.

but ye ideally code should b able to handle all combinations of existing hash/ existing url or not, and existing actual file or not